sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/InMemoryPartitionManager.cs

```diff
@@ -0,0 +1,140 @@
+
+        /// <summary>The set of stored ownership.  Partition ids are used as keys.</summary>
+        private Dictionary<string, PartitionOwnership> Ownership;
+
+        /// <summary>Logs activities performed by this partition manager.</summary>
+        private Action<string> Logger;
+
+        /// <summary>
+        ///   Initializes a new instance of the <see cref="InMemoryPartitionManager"/> class.
+        /// </summary>
+        ///
+        /// <param name="logger">Logs activities performed by this partition manager.</param>
+        ///
+        public InMemoryPartitionManager(Action<string> logger = null)
+        {
+            Logger = logger;
+
+            Checkpoints = new ConcurrentDictionary<string, Checkpoint>();
+            Ownership = new Dictionary<string, PartitionOwnership>();
+        }
+
+        /// <summary>
+        ///   Retrieves a complete ownership list from the in-memory storage service.
+        /// </summary>
+        ///
+        /// <param name="eventHubName">The name of the specific Event Hub the ownership are associated with, relative to the Event Hubs namespace that contains it.</param>
+        /// <param name="consumerGroup">The name of the consumer group the ownership are associated with.</param>
+        ///
+        /// <returns>An enumerable containing all the existing ownership for the associated Event Hub and consumer group.</returns>
+        ///
+        public override Task<IEnumerable<PartitionOwnership>> ListOwnershipAsync(string eventHubName,
+                                                                                 string consumerGroup)
+        {
+            lock(OwnershipClaimLock)
+            {
+                var ownershipList = Ownership.Values
+                    .Where(ownership => ownership.EventHubName == eventHubName &&
+                        ownership.ConsumerGroup == consumerGroup)
+                    .ToList();
+
+                return Task.FromResult((IEnumerable<PartitionOwnership>)ownershipList);
+            }
+        }
+
+        /// <summary>
+        ///   Tries to claim a list of specified ownership.
+        /// </summary>
+        ///
+        /// <param name="partitionOwnership">An enumerable containing all the ownership to claim.</param>
+        ///
+        /// <returns>An enumerable containing the successfully claimed ownership.</returns>
+        ///
+        public override Task<IEnumerable<PartitionOwnership>> ClaimOwnershipAsync(IEnumerable<PartitionOwnership> partitionOwnership)
+        {
+            var claimedOwnership = new List<PartitionOwnership>();
+
+            foreach(var ownership in partitionOwnership)
+            {
+                // The following lock makes sure two different event processors won't try to claim ownership of a partition
+                // simultaneously.  This approach prevents an ownership from being stolen just after being claimed.
+
+                lock (OwnershipClaimLock)
```

Comment:
I'd  recommend locking outside the loop, for what its worth;  that way nothing else can steal the lock during the ownership claim and you don't pay a heavier cost for acquiring the lock multiple times.   Not a blocker, but probably worth a quick update.

Reply:
Done.